### PR TITLE
feat(anta.cli): Check catalog functionality

### LINKED ISSUE #407: Optimize runner once #377 is pushed
Depends on #377

Cartesian product of device x test
* make device optional in test to instantiate it to check on filters per device afterwards
* make device mandatory to run AntaTest.test

### LINKED ISSUE #408: Feat(anta.cli): Implement anta check catalog / anta check inventory
* for catalog - probably depends on #407 (possible without it with mock device)

### PR DESCRIPTION
**Pull request by gmuloc**
# Description

- Add `anta check catalog -c <CATALOG>`
- Add `AntaCatalog` class to manipulate test catalog
   - `tests` property can be used to get and set tests to the catalog.
   - `get_tests_by_tags` to filter tests by tag
   - `get_tests_by_device` to filter tests by device
- Add `AntaCatalogFile` pydantic model to provide validation for test catalog YAML file
    ``` python
    catalog = AntaCatalog(filename='/mnt/lab/my_anta_catalog.yaml')
    catalog.check()
    ```
- Add `AntaTestDefinition` pydantic model to provide validation mainly for programmatic `AntaCatalog` instantiation.
    ``` python
    AntaCatalog(tests=[(AntaTest, AntaTest.Input())])
    ```

Fixes #408
Fixes #407 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have run pre-commit for code linting and typing (`pre-commit run`)
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes (`tox -e testenv`)

### CODE REVIEW

File: anta/catalog.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        mock_device = AsyncEOSDevice(name="mock", host="127.0.0.1", username="mock", password="mock")
+
+        manager = ResultManager()
+        # Instantiate each test to verify the Inputs are correct
+        for test_class, test_inputs in self.tests:
+            # TODO - this is the same code with typing as in runner.py but somehow mypy complains that test_class
+            # ot type AntaTest is not callable
+            test_instance = test_class(device=mock_device, inputs=test_inputs)  # type: ignore[operator]
+            manager.add_test_result(test_instance.result)
+        return manager
```
Comment (gmuloc) [inline anta/catalog.py:152]: this is crap use https://docs.pydantic.dev/latest/concepts/models/#helper-functions `model_validate`

File: .vscode/settings.json [hunk 2]
```diff
@@ -1,23 +1,22 @@
...
+  "mypy-type-checker.importStrategy": "fromEnvironment",
+  "mypy-type-checker.args": [
     "--config-file=pyproject.toml"
   ],
-  "pylint.severity": {
-    "refactor": "Warning"
+  "isort.importStrategy": "fromEnvironment",
+  "isort.check": true,
+  "files.associations": {
+    "*.yaml": "home-assistant"
```
Comment (gmuloc) [inline .vscode/settings.json:20]: so you are using home-assistant with ANTA ? :p

Comment (mtache) [reply .vscode/settings.json:20]: f**k I am fighting with this extension

**Review by mtache — APPROVED** [review]

File: anta/catalog.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        self._tests = value
+
+    def _parse_file(self) -> None:
+        """
+        Parse the catalog YAML file
+        """
+        if self.filename:
+            try:
+                with open(file=self.filename, mode="r", encoding="UTF-8") as file:
+                    self._data = safe_load(file)
```
Comment (gmuloc) [inline anta/catalog.py:250]: lets put this `self.tests`

**Review by carl-baillargeon — COMMENTED** [review]
I think the `WARNING` log should be in `DEBUG`
```bash
➜  anta git:(issue/408) anta nrfu table            
Catalog /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml is valid
╭────────────────────── Settings ──────────────────────╮
│ Running ANTA tests:                                  │
│ - ANTA Inventory contains 1 devices (AsyncEOSDevice) │
│ - Tests catalog contains 99 tests                    │
╰──────────────────────────────────────────────────────╯

[09:41:15] WARNING  Overriding AntaCatalog data from file /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml 
```

**Review by carl-baillargeon — COMMENTED** [review]
Investigate the option of having a single `ANTA_CATALOG` env for both `anta nrfu` and `anta check`.

Also `ANTA_CHECK_CATALOG_CATALOG` is weird.
```bash
➜  anta git:(issue/408) anta check catalog --help
Usage: anta check catalog [OPTIONS]

  Check that the catalog is valid

Options:
  -c, --catalog FILE  Path to the tests catalog YAML file  [env var:
                      ANTA_CHECK_CATALOG_CATALOG; required]
  --help              Show this message and exit.
  ```

**Review by carl-baillargeon — COMMENTED** [review]
When using the device filter, I would expect the test to run only on the specified devices.
The following example shows the test being run on DC1-LEAF1A even though the filter is for DC1-LEAF1B

Catalog:
```yaml
---
anta.tests.aaa:
  - VerifyTacacsSourceIntf:
      intf: Loopback0
      vrf: default
      filters:
        devices:
          - DC1-LEAF1B
  ```
  Running anta:
  ```bash
➜  anta git:(issue/408) anta nrfu text 
Catalog /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml is valid
╭────────────────────── Settings ──────────────────────╮
│ Running ANTA tests:                                  │
│ - ANTA Inventory contains 1 devices (AsyncEOSDevice) │
│ - Tests catalog contains 1 tests                     │
╰──────────────────────────────────────────────────────╯

[10:26:34] WARNING  Overriding AntaCatalog data from file                                                                                  catalog.py:264
                    /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml                                                  
           INFO     Running ANTA tests...                                                                                                   runner.py:100
[10:26:35] INFO     Cache statistics for DC1-LEAF1A: {'total_commands_sent': 1, 'cache_hits': 0, 'cache_hit_ratio': '0.00%'}                runner.py:113
  • Running NRFU Tests...100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1/1 • 0:00:00 • 0:00:00

DC1-LEAF1A :: VerifyTacacsSourceIntf :: FAILURE (Source-interface Loopback0 is not configured in VRF default)                        
```
Also if the device filter matches a device, the test is run twice:

Catalog:
```yaml
---
anta.tests.aaa:
  - VerifyTacacsSourceIntf:
      intf: Loopback0
      vrf: default
      filters:
        devices:
          - DC1-LEAF1A
```
Running anta:
```bash
➜  anta git:(issue/408) anta nrfu text 
Catalog /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml is valid
╭────────────────────── Settings ──────────────────────╮
│ Running ANTA tests:                                  │
│ - ANTA Inventory contains 1 devices (AsyncEOSDevice) │
│ - Tests catalog contains 1 tests                     │
╰──────────────────────────────────────────────────────╯

[10:31:08] WARNING  Overriding AntaCatalog data from file                                                                             catalog.py:264
                    /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml                                             
           INFO     Running ANTA tests...                                                                                              runner.py:100
           INFO     Cache statistics for DC1-LEAF1A: {'total_commands_sent': 2, 'cache_hits': 1, 'cache_hit_ratio': '50.00%'}          runner.py:113
  • Running NRFU Tests...100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 • 0:00:00 • 0:00:00

DC1-LEAF1A :: VerifyTacacsSourceIntf :: FAILURE (Source-interface Loopback0 is not configured in VRF default)
DC1-LEAF1A :: VerifyTacacsSourceIntf :: FAILURE (Source-interface Loopback0 is not configured in VRF default)
```

**Review by carl-baillargeon — COMMENTED** [review]
`Catalog None is valid` is not necessary here.
```bash
➜  anta git:(issue/408) anta nrfu text --help
Catalog None is valid
```

**Review by carl-baillargeon — COMMENTED** [review]
When using tags that match a device tag but 0 tests, it would be nice to have an INFO log and exit, similar to when we use a tag that do not match any device:

No logs and not exiting:

```bash
➜  anta git:(issue/408) anta nrfu text --tags DC1       
Catalog /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml is valid
╭────────────────────── Settings ──────────────────────╮
│ Running ANTA tests:                                  │
│ - ANTA Inventory contains 1 devices (AsyncEOSDevice) │
│ - Tests catalog contains 1 tests                     │
╰──────────────────────────────────────────────────────╯

[10:36:54] WARNING  Overriding AntaCatalog data from file /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml                                                                                                                                     catalog.py:264
           INFO     Running ANTA tests...                                                                                                                                                                                                                                          runner.py:100
           INFO     Cache statistics for DC1-LEAF1A: {'total_commands_sent': 0, 'cache_hits': 0, 'cache_hit_ratio': '0.00%'}
```
Log + exiting:
```bash
➜  anta git:(issue/408) anta nrfu text --tags AAA
Catalog /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml is valid
╭────────────────────── Settings ──────────────────────╮
│ Running ANTA tests:                                  │
│ - ANTA Inventory contains 1 devices (AsyncEOSDevice) │
│ - Tests catalog contains 1 tests                     │
╰──────────────────────────────────────────────────────╯

[10:36:39] WARNING  Overriding AntaCatalog data from file /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml                                                                                                                                     catalog.py:264
           INFO     No device in the established state 'True' matching the tags ['AAA'] was found. There is no device to run tests against, exiting  
```

File: anta/catalog.py (hunk 3 above)
Comment (gmuloc) [reply anta/catalog.py:250]: this means no need for check anymore

**Comment by gmuloc** [comment]
> I think the `WARNING` log should be in `DEBUG`
> 
> ```shell
> ➜  anta git:(issue/408) anta nrfu table            
> Catalog /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml is valid
> ╭────────────────────── Settings ──────────────────────╮
> │ Running ANTA tests:                                  │
> │ - ANTA Inventory contains 1 devices (AsyncEOSDevice) │
> │ - Tests catalog contains 99 tests                    │
> ╰──────────────────────────────────────────────────────╯
> 
> [09:41:15] WARNING  Overriding AntaCatalog data from file /home/cbaillar/git_projects/ansible-avd-examples/multidomain-evpn/anta_catalog.yaml 
> ```

I think it should probably be warning - the more interesting question is why you are seeing this as self._tests should be empty in your case

**Comment by gmuloc** [comment]
> Investigate the option of having a single `ANTA_CATALOG` env for both `anta nrfu` and `anta check`.
> 
> Also `ANTA_CHECK_CATALOG_CATALOG` is weird.
> 
> ```shell
> ➜  anta git:(issue/408) anta check catalog --help
> Usage: anta check catalog [OPTIONS]
> 
>   Check that the catalog is valid
> 
> Options:
>   -c, --catalog FILE  Path to the tests catalog YAML file  [env var:
>                       ANTA_CHECK_CATALOG_CATALOG; required]
>   --help              Show this message and exit.
> ```

Fixed in [2c39acd](https://github.com/arista-netdevops-community/anta/pull/414/commits/2c39acdc6f4c3014de23c2130e0ad23770fbe649)

File: anta/catalog.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+            context={"test": data["test"]},
+        )
+        super(BaseModel, self).__init__()
+
+    @model_serializer
+    def ser_model(self) -> Dict[str, AntaTest.Input]:
+        """
+        Serialize an AntaTestDefinition as it is defined in a test catalog YAML file.
+        """
+        return {self.test.__name__: self.inputs}
```
Comment (carl-baillargeon) [inline anta/catalog.py:54]: `self.inputs` should be serialized too no?

Comment (mtache) [reply anta/catalog.py:54]: Yes, actually this method is not used or tested yet.
The idea is to be able to dump a YAML file from an `AntaCatalog` object.
The output we get from `anta check catalog` comes from the __repr__() and is not a serialisation.

So we either delete it or leave it as is with a TODO.

File: anta/logger.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
+# Copyright (c) 2023 Arista Networks, Inc.
+# Use of this source code is governed by the Apache License 2.0
+# that can be found in the LICENSE file.
+"""
+Loader that parses a YAML test catalog and imports corresponding Python functions
```
Comment (carl-baillargeon) [inline anta/logger.py:5]:
```suggestion
Configure logging for ANTA
```

**Review by carl-baillargeon — APPROVED** [review]
LGTM. Tested with AVD as well. Thanks!!

**Pull request merged**